cluster-api/src/main/java/io/aiven/klaw/clusterapi/services/ConsumerGroupService.java

```diff
@@ -0,0 +1,257 @@
+      offsetAndMetadataMap.put(topicPartition, new OffsetAndMetadata(partitionOffset));
+    }
+
+    return offsetAndMetadataMap;
+  }
+
+  public Map<String, Long> getCurrentOffsetsPositions(String consumerGroup, AdminClient adminClient)
+      throws Exception {
+    Map<String, Long> currentOffsetPositionsMap = new TreeMap<>();
+    KafkaFuture<Map<TopicPartition, OffsetAndMetadata>> offsetsBeforeResetFuture =
+        adminClient.listConsumerGroupOffsets(consumerGroup).partitionsToOffsetAndMetadata();
+    try {
+      Map<TopicPartition, OffsetAndMetadata> topicPartitionOffsetAndMetadataMap =
+          offsetsBeforeResetFuture.get();
+      for (TopicPartition topicPartition : topicPartitionOffsetAndMetadataMap.keySet()) {
+        currentOffsetPositionsMap.put(
+            topicPartition.toString(),
+            topicPartitionOffsetAndMetadataMap.get(topicPartition).offset());
+      }
+      return currentOffsetPositionsMap;
+    } catch (InterruptedException | ExecutionException e) {
+      // ignore error as there may not be any events or couldn't retrieve events
+      return new HashMap<>();
+    }
+  }
+
+  public TopicDescription describeTopic(AdminClient adminClient, String topicName)
+      throws Exception {
+    try {
+      DescribeTopicsResult result = adminClient.describeTopics(Collections.singleton(topicName));
+      if (result.values().containsKey(topicName)) {
+        return result.values().get(topicName).get();
+      } else {
+        return null;
+      }
+    } catch (Exception ex) {
+      if (ex.getLocalizedMessage().contains("UnknownTopicOrPartitionException")) {
+        log.info("Topic {} does not exist on the cluster.", topicName);
+        return null;
+      } else {
+        String errorMessage = "Unable to describe the topic " + topicName;
+        log.error(errorMessage, ex);
+        throw new Exception(errorMessage, ex);
+      }
+    }
+  }
+
+  public List<OffsetDetails> getConsumerGroupDetails(
+      String consumerGroupId,
+      String topicName,
+      String environment,
+      KafkaSupportedProtocol protocol,
+      String clusterName)
+      throws Exception {
+    log.info(
+        "getConsumerGroupDetails topicName {} consumerGroupId {} environment {} protocol {} clusterName {}",
+        topicName,
+        consumerGroupId,
+        environment,
+        protocol,
+        clusterName);
+
+    AdminClient adminClient = clusterApiUtils.getAdminClient(environment, protocol, clusterName);
+
+    List<OffsetDetails> consumerGroupOffsetList = new ArrayList<>();
+    OffsetDetails offsetDetails;
+    try {
+      DescribeTopicsResult describeTopicsResult =
+          adminClient.describeTopics(Collections.singletonList(topicName));
+      List<TopicPartitionInfo> topicPartitions =
+          describeTopicsResult.values().get(topicName).get().partitions();
+
+      TopicPartition topicPartition;
+      Map<TopicPartition, OffsetSpec> topicPartitionOffsetSpecMap = new HashMap<>();
+
```

Comment:
I think the two for loops below can be pulled into a method with the OffsetSpec.earliest() and .latest as variables returning the the ListOffsetResult

Reply:
updated